File: internal/parser/parser.go
package parser

import (
	"bufio"
	"encoding/json"
	"fmt"
	"os"
	"strings"
)

var validInstructions = map[string]bool{
	"FROM": true, "COPY": true, "RUN": true,
	"WORKDIR": true, "ENV": true, "CMD": true,
}

// Instruction represents one parsed line from a Docksmithfile.
type Instruction struct {
	Type       string
	Args       string
	LineNumber int
}

// ParseDocksmithfile reads a Docksmithfile and returns all instructions.
func ParseDocksmithfile(filePath string) ([]Instruction, error) {
	f, err := os.Open(filePath)
	if err != nil {
		return nil, fmt.Errorf("[BUILD ERROR] Docksmithfile not found at: %s", filePath)
	}
	defer f.Close()

	var instructions []Instruction
	scanner := bufio.NewScanner(f)
	lineNum := 0

	for scanner.Scan() {
		lineNum++
		line := strings.TrimSpace(scanner.Text())

		if line == "" || strings.HasPrefix(line, "#") {
			continue
		}

		parts := strings.SplitN(line, " ", 2)
		keyword := strings.ToUpper(parts[0])
		args := ""
		if len(parts) > 1 {
			args = strings.TrimSpace(parts[1])
		}

		if !validInstructions[keyword] {
			return nil, fmt.Errorf("[PARSE ERROR] Line %d: Unknown instruction %q.\n  Valid instructions: CMD, COPY, ENV, FROM, RUN, WORKDIR\n  Got: %q",
				lineNum, parts[0], line)
		}

		if err := validateArgs(keyword, args, lineNum); err != nil {
			return nil, err
		}

		instructions = append(instructions, Instruction{
			Type:       keyword,
			Args:       args,
			LineNumber: lineNum,
		})
	}

	if len(instructions) == 0 {
		return nil, fmt.Errorf("[PARSE ERROR] Docksmithfile is empty or has no valid instructions")
	}
	if instructions[0].Type != "FROM" {
		return nil, fmt.Errorf("[PARSE ERROR] Line %d: Docksmithfile must start with FROM, got %q",
			instructions[0].LineNumber, instructions[0].Type)
	}

	return instructions, nil
}

func validateArgs(keyword, args string, lineNum int) error {
	switch keyword {
	case "FROM":
		if args == "" {
			return fmt.Errorf("[PARSE ERROR] Line %d: FROM needs an image name.\n  Example: FROM alpine:latest", lineNum)
		}
	case "COPY":
		if len(strings.Fields(args)) < 2 {
			return fmt.Errorf("[PARSE ERROR] Line %d: COPY needs <src> and <dest>.\n  Example: COPY . /app", lineNum)
		}
	case "RUN":
		if args == "" {
			return fmt.Errorf("[PARSE ERROR] Line %d: RUN needs a command.\n  Example: RUN echo hello", lineNum)
		}
	case "WORKDIR":
		if args == "" {
			return fmt.Errorf("[PARSE ERROR] Line %d: WORKDIR needs a path.\n  Example: WORKDIR /app", lineNum)
		}
	case "ENV":
		if !strings.Contains(args, "=") {
			return fmt.Errorf("[PARSE ERROR] Line %d: ENV must be KEY=value.\n  Example: ENV APP_NAME=myapp\n  Got: %q", lineNum, args)
		}
	case "CMD":
		var result []interface{}
		if err := json.Unmarshal([]byte(args), &result); err != nil {
			return fmt.Errorf("[PARSE ERROR] Line %d: CMD must be a JSON string array.\n  Example: CMD [\"python\", \"main.py\"]\n  Got: %q", lineNum, args)
		}
		for _, v := range result {
			if _, ok := v.(string); !ok {
				return fmt.Errorf("[PARSE ERROR] Line %d: CMD array must contain only strings", lineNum)
			}
		}
	}
	return nil
}

// ParseFromArgs parses "alpine:latest" → ("alpine", "latest").
func ParseFromArgs(args string) (name, tag string) {
	if idx := strings.Index(args, ":"); idx >= 0 {
		return strings.TrimSpace(args[:idx]), strings.TrimSpace(args[idx+1:])
	}
	return strings.TrimSpace(args), "latest"
}

// ParseEnvArgs parses "KEY=value" → ("KEY", "value").
func ParseEnvArgs(args string) (key, value string, err error) {
	idx := strings.Index(args, "=")
	if idx < 0 {
		return "", "", fmt.Errorf("[PARSE ERROR] ENV must be KEY=value. Got: %q", args)
	}
	return strings.TrimSpace(args[:idx]), strings.TrimSpace(args[idx+1:]), nil
}

// ParseCopyArgs parses ". /app" → (".", "/app").
func ParseCopyArgs(args string) (src, dest string, err error) {
	parts := strings.Fields(args)
	if len(parts) < 2 {
		return "", "", fmt.Errorf("[PARSE ERROR] COPY needs <src> and <dest>. Got: %q", args)
	}
	return parts[0], parts[1], nil
}

// ParseCmdArgs parses a JSON array string into []string.
func ParseCmdArgs(args string) ([]string, error) {
	var result []string
	if err := json.Unmarshal([]byte(args), &result); err != nil {
		return nil, fmt.Errorf("[PARSE ERROR] CMD must be a JSON array. Got: %q", args)
	}
	return result, nil
}

File: docksmith/parser.py
# docksmith/parser.py
# ============================================================
#  PROTHAM — File 1: Docksmithfile Parser
#  Test: python3 -m pytest tests/test_parser.py -v
# ============================================================

import json
from dataclasses import dataclass

VALID_INSTRUCTIONS = {"FROM", "COPY", "RUN", "WORKDIR", "ENV", "CMD"}


@dataclass
class Instruction:
    type: str         # FROM, COPY, RUN, WORKDIR, ENV, CMD
    args: str         # Everything after the keyword
    line_number: int  # Original line number (for error messages)

    def __repr__(self):
        return f"Instruction(type={self.type!r}, args={self.args!r}, line={self.line_number})"


# ── Main parser ───────────────────────────────────────────────────────────────

def parse_docksmithfile(filepath: str) -> list:
    """
    Reads a Docksmithfile and returns a list of Instruction objects.
    Fails immediately with clear error + line number on unknown instructions.
    """
    instructions = []

    try:
        with open(filepath, "r") as f:
            lines = f.readlines()
    except FileNotFoundError:
        raise FileNotFoundError(
            f"[BUILD ERROR] Docksmithfile not found at: {filepath}"
        )

    for i, raw_line in enumerate(lines, start=1):
        line = raw_line.strip()

        # Skip blank lines and comments
        if not line or line.startswith("#"):
            continue

        # Split keyword from rest of line
        parts = line.split(None, 1)
        keyword = parts[0].upper()
        args = parts[1].strip() if len(parts) > 1 else ""

        # Fail immediately on unknown instruction
        if keyword not in VALID_INSTRUCTIONS:
            raise ValueError(
                f"[PARSE ERROR] Line {i}: Unknown instruction '{parts[0]}'.\n"
                f"  Valid instructions: {', '.join(sorted(VALID_INSTRUCTIONS))}\n"
                f"  Got: {line!r}"
            )

        # Validate args for each instruction
        _validate_args(keyword, args, i)

        instructions.append(Instruction(type=keyword, args=args, line_number=i))

    if not instructions:
        raise ValueError(
            f"[PARSE ERROR] Docksmithfile is empty or has no valid instructions."
        )

    # Must start with FROM
    if instructions[0].type != "FROM":
        raise ValueError(
            f"[PARSE ERROR] Line {instructions[0].line_number}: "
            f"Docksmithfile must start with FROM, got '{instructions[0].type}'"
        )

    return instructions


def _validate_args(keyword: str, args: str, line_number: int):
    """Validates each instruction has correct argument format."""

    if keyword == "FROM":
        if not args:
            raise ValueError(
                f"[PARSE ERROR] Line {line_number}: FROM needs an image name.\n"
                f"  Example: FROM alpine:latest"
            )

    elif keyword == "COPY":
        if len(args.split(None, 1)) < 2:
            raise ValueError(
                f"[PARSE ERROR] Line {line_number}: COPY needs <src> and <dest>.\n"
                f"  Example: COPY . /app"
            )

    elif keyword == "RUN":
        if not args:
            raise ValueError(
                f"[PARSE ERROR] Line {line_number}: RUN needs a command.\n"
                f"  Example: RUN echo hello"
            )

    elif keyword == "WORKDIR":
        if not args:
            raise ValueError(
                f"[PARSE ERROR] Line {line_number}: WORKDIR needs a path.\n"
                f"  Example: WORKDIR /app"
            )

    elif keyword == "ENV":
        if "=" not in args:
            raise ValueError(
                f"[PARSE ERROR] Line {line_number}: ENV must be KEY=value.\n"
                f"  Example: ENV APP_NAME=myapp\n"
                f"  Got: {args!r}"
            )

    elif keyword == "CMD":
        try:
            result = json.loads(args)
            if not isinstance(result, list):
                raise ValueError()
            if not all(isinstance(x, str) for x in result):
                raise ValueError()
        except (json.JSONDecodeError, ValueError):
            raise ValueError(
                f"[PARSE ERROR] Line {line_number}: CMD must be a JSON string array.\n"
                f"  Example: CMD [\"python\", \"main.py\"]\n"
                f"  Got: {args!r}"
            )


# ── Argument parsers (called by builder.py) ───────────────────────────────────

def parse_from_args(args: str) -> tuple:
    """
    Parses 'alpine:latest' → ('alpine', 'latest')
    Defaults tag to 'latest' if not specified.
    """
    if ":" in args:
        name, tag = args.split(":", 1)
    else:
        name, tag = args, "latest"
    return name.strip(), tag.strip()


def parse_env_args(args: str) -> tuple:
    """
    Parses 'KEY=value' → ('KEY', 'value')
    Handles values that contain = signs (e.g. URL=http://x.com/a=b)
    """
    if "=" not in args:
        raise ValueError(f"[PARSE ERROR] ENV must be KEY=value format. Got: {args!r}")
    key, _, value = args.partition("=")
    return key.strip(), value.strip()


def parse_copy_args(args: str) -> tuple:
    """
    Parses '. /app' → ('.', '/app')
    """
    parts = args.split(None, 1)
    if len(parts) < 2:
        raise ValueError(
            f"[PARSE ERROR] COPY needs <src> and <dest>. Got: {args!r}"
        )
    return parts[0], parts[1]


def parse_cmd_args(args: str) -> list:
    """
    Parses '["python", "main.py"]' → ['python', 'main.py']
    """
    try:
        result = json.loads(args)
        if not isinstance(result, list):
            raise ValueError()
        return result
    except (json.JSONDecodeError, ValueError):
        raise ValueError(
            f"[PARSE ERROR] CMD must be a JSON array.\n"
            f"  Example: CMD [\"python\", \"main.py\"]\n"
            f"  Got: {args!r}"
        )